sklearn/preprocessing/_polynomial.py

```diff
@@ -23,6 +23,32 @@
 ]
 
 
+def _cast_to_int64_if_needed(X, degree, interaction_only):
+    """Computes the expanded dimensionality and casts X to int64 if the
+    expanded dim is too big"""
+    d = int(X.shape[1])
+    assert degree in (2, 3)
+    if degree == 2:
+        expanded_dimensionality = (d ** 2 + d) / 2 - interaction_only * d
+    else:
+        expanded_dimensionality = ((d ** 3 + 3 * d ** 2 + 2 * d) / 6 -
+                                   interaction_only * d ** 2)
+    if expanded_dimensionality > np.iinfo(np.int64).max:
+        raise ValueError("The expanded dimensionality will be too big to "
+                         "be contained in an int64.")
+    if expanded_dimensionality > np.iinfo(np.int32).max:
+        # if the expansion needs int64 ints, we cast every index value in X
+        # to int64
+        X = X.copy()
+        X.data = X.data.astype(np.int64)
+        X.indices = X.indices.astype(np.int64)
+        X.indptr = X.indptr.astype(np.int64)
```

Comment:
There are two copies being created here, one from `X.copy` and here when changing the dtype of the indices. I do not see a good alternative. I would expect the following to work:

```python
        X = csr_matrix(
            (X.data, X.indices.astype(np.int64), X.indptr.astype(np.int64)),
            shape=X.shape, copy=True,
        )
```

but internally `csr_matrix` checks the indices and will cast back down to `np.int32`.

Reply:
That's right, your solution sounds good
I'm not completely sure to understand the problem with it: is it because when creating a `csr_matrix` with indices that could hold in `int32`, scipy automatically choses `int32` indices ? 

In this case maybe I could avoid copying `X`, but deal with every part of it (data, indices, inptr) separately, like:
- remove the `X = X.copy()` line
- do `X_indices = X.indices.astype(np.int64, copy=False)` (to do a copy only if the type is changed)
- similarly `X_indptr =X_indptr.astype(np.int64, copy=False)`
And then return `X.data`, and `X_indices` and `X_indptr`, which I would use later on in `_csr_polynomial_expansion`
What do you think ?